pkg/operator/clustermembercontroller2/clustermembercontroller.go

```diff
@@ -0,0 +1,382 @@
+		return []string{}, fmt.Errorf("could not find etcd address in host-etcd")
+	}
+	var endpoints []string
+	for _, addr := range hostEtcd.Subsets[0].Addresses {
+		endpoints = append(endpoints, fmt.Sprintf("https://%s.%s:2379", addr.Hostname, c.etcdDiscoveryDomain))
+	}
+	return endpoints, nil
+}
+
+func (c *ClusterMemberController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(workQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(workQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(workQueueKey) },
+	}
+}
+
+func (c *ClusterMemberController) isEtcdHealthy() (bool, error) {
+	// getting a new client everytime because we dont know what etcd-membership looks like
+	etcdClient, err := c.getEtcdClient()
+	defer etcdClient.Close()
+	if err != nil {
+		c.eventRecorder.Warningf("ErrorGettingEtcdClient", "error getting etcd client: %#v", err)
+		return false, err
+	}
+
+	memberList, err := etcdClient.MemberList(context.Background())
+	if err != nil {
+		c.eventRecorder.Warningf("ErrorGettingMemberList", "error getting etcd member list: %#v", err)
+		return false, err
+	}
+	for _, member := range memberList.Members {
+		statusResp, err := etcdClient.Status(context.Background(), member.ClientURLs[0])
+		if err != nil {
+			c.eventRecorder.Warningf("EtcdMemberNotHealthy", "etcd member %s is not healthy: %#v", member.Name, err)
+			// since error is indicative of unhealthy member, not returning
+			// the actual error
+			return false, nil
+		}
+		klog.V(4).Infof("etcd member %s is healthy committed and with %s index", member.Name, statusResp.RaftIndex)
+	}
+	return true, nil
+}
+
+func (c *ClusterMemberController) getPeerToScale() (string, []error) {
+	// list etcd member pods
+	etcdPodRequirement, err := labels.NewRequirement("k8s-app", selection.In, []string{"etcd"})
+	if err != nil {
+		c.eventRecorder.Warningf("ErrorGettingLabelSelector", "error get k8s-app=etcd selector: %#v", err)
+		return "", []error{err}
+	}
+	etcdPodSelector := labels.NewSelector().Add(*etcdPodRequirement)
+	pods, err := c.podLister.List(etcdPodSelector)
+	if err != nil {
+		c.eventRecorder.Warningf("ErrorListingPods", "error listing pods with label selector k8s-app=etcd: %#v", err)
+		return "", []error{err}
+	}
+
+	// go through the list of all pods, pick one peerFQDN to return from unready pods
+	// and collect dns resolution errors on the way.
+	var dnsErrors []error
+	peerFQDN := ""
+	for _, pod := range pods {
+		ready := false
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == corev1.PodReady {
+				ready = condition.Status == corev1.ConditionTrue
+				break
+			}
+		}
+		if !ready {
+			c.eventRecorder.Eventf("FoundPodToScale", "found pod %s to scale in etcd membership", pod.Name)
+			peerFQDN, err = clustermembercontroller.ReverseLookupSelf("etcd-server-ssl", "tcp", c.etcdDiscoveryDomain, pod.Status.HostIP)
```

Comment:
no, we don't base anything on pod IPs.  Use the node's internal IP address.

Reply:
sure, fixing it